irohad/ordering/impl/ordering_gate_cache/on_demand_resend_strategy.cpp

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "ordering/impl/ordering_gate_cache/on_demand_resend_strategy.hpp"
+
+#include <boost/assert.hpp>
+#include <iostream>
+#include "interfaces/iroha_internal/transaction_batch.hpp"
+#include "interfaces/transaction.hpp"
+#include "ordering/impl/on_demand_common.hpp"
+
+using namespace iroha::ordering;
+
+bool OnDemandResendStrategy::feed(
+    std::shared_ptr<shared_model::interface::TransactionBatch> batch) {
+  std::shared_lock<std::shared_timed_mutex> lock(access_mutex_);
+  auto res =
+      sent_batches_.emplace(batch, std::make_pair(current_round_, false));
+  return res.second;
+}
+
+bool OnDemandResendStrategy::readyToUse(
+    std::shared_ptr<shared_model::interface::TransactionBatch> batch) {
+  std::shared_lock<std::shared_timed_mutex> lock(access_mutex_);
+  auto batch_found = std::find_if(
+      sent_batches_.begin(), sent_batches_.end(), [&batch](auto saved_batch) {
+        return *(saved_batch.first) == *batch;
+      });
+  if (batch_found != sent_batches_.end()) {
+    batch_found->second.second = true;
+    return true;
+  }
+  return false;
+}
+
+OnDemandResendStrategy::RoundSetType OnDemandResendStrategy::extract(
+    std::shared_ptr<shared_model::interface::TransactionBatch> batch) {
+  std::shared_lock<std::shared_timed_mutex> lock(access_mutex_);
+  std::set<consensus::Round> valid_rounds{
+      nextCommitRound(nextCommitRound(current_round_)),
+      nextCommitRound(nextRejectRound(current_round_)),
+      nextRejectRound(nextCommitRound(current_round_)),
+      nextRejectRound(nextRejectRound(current_round_))};
+
+  auto saved_round_iterator = std::find_if(
+      sent_batches_.begin(), sent_batches_.end(), [&batch](auto saved_batch) {
+        return *(saved_batch.first) == *batch;
+      });
+  if (saved_round_iterator == sent_batches_.end()) {
+    return valid_rounds;
+  }
+  if (not saved_round_iterator->second.second) {
+    return valid_rounds;
+  }
+  auto saved_round = saved_round_iterator->second.first;
+
+  if (nextCommitRound(nextCommitRound(saved_round)) == current_round_) {
```

Comment:
Could this whole block
https://github.com/hyperledger/iroha/blob/e0dcced4843337d48dac7f269aaac21830ca9515/irohad/ordering/impl/ordering_gate_cache/on_demand_resend_strategy.cpp#L59-L73
be substituted with something like
```c++
(pseudo-code)
std::set<consensus::Round> target_rounds;
copy_if(valid_rounds,
    inserter(target_rounds),
    [current_round_](target_round){ return target_round.block_round <= current_round_.block_round; });
```

Reply:
I afraid it will work only for (reject, reject) case.

Reply:
yes, I understood your code wrong way. as we discussed, could you apply something like
```
target_rounds = setDifference(
    reachableIn2Steps(current_round_),
    reachableIn2Steps(saved_round));
```